### PR TITLE
Feature/wmt 1028/input validation

### PR DESCRIPTION
**Pull request by tswann**
- Apply server-side validation to all input forms (contracted hours, reductions, workload points)
- Add GOV.UK style form validation error messages
- FieldValidator and FieldSetValidator in `services/validators`
- ValidationError class in `services/errors`
- New domain object isValid() method for Reductions
- New WorkloadPoints domain object
- Update all tests with new validation approach

### CODE REVIEW

File: app/services/domain/workload-points.js [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      .isFloat(0, 100.0)
+    FieldValidator(this.defaultContractedHoursPo, 'defaultContractedHoursPo', errors)
+      .isRequired()
+      .isInt(0, 37)
+    FieldValidator(this.defaultContractedHoursPso, 'defaultContractedHoursPso', errors)
+      .isRequired()
+      .isInt(0, 37)
+    FieldValidator(this.parom, 'parom', errors)
+      .isRequired()
+      .isInt(0, 999)
```
Comment (caoimheb) [inline app/services/domain/workload-points.js]: Again, are these all agreed boundaries?

File: app/views/workload-points.html [hunk 2]
```diff
@@ -41,6 +41,9 @@
                 <tbody>
                   <tr>
```
Comment (caoimheb) [inline app/views/workload-points.html]: Could we include a check on this page that if there are errors the form is already in edit mode?

File: app/services/validators/validation-error-messages.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
+  getIsRequired: function (displayName) { return `${displayName} is required` },
+  getIsAlpha: function (displayName) { return `${displayName} must only contain letters` },
+  getIsNumeric: function (displayName) { return `${displayName} must only contain numbers` },
```
Comment (caoimheb) [inline app/services/validators/validation-error-messages.js]: Naming - Add `Message` to the end of these first three?

File: package.json [hunk 4]
```diff
@@ -39,11 +39,11 @@
     "gulp-webdriver": "^2.0.3",
     "mocha": "3.2.x",
     "proxyquire": "1.7.x",
+    "selenium-standalone": "6.5.0",
```
Comment (caoimheb) [inline package.json]: 👍

File: app/services/validators/validation-field-names.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
+module.exports = {
+  'hours': 'Contracted hours',
+  'reasonForReductionId': 'Reason for reduction',
+  'reduction-hours': 'Reduction hours',
+  'reduction_start_date': 'Reduction start date',
+  'reduction_end_date': 'Reduction end date',
```
Comment (caoimheb) [inline app/services/validators/validation-field-names.js]: Three different naming conventions on one form? 🙁  ^^

Comment (caoimheb) [inline app/services/validators/validation-field-names.js]: Field naming missing for `notes`. Error coming through as `undefined must be shorter than 4000 characters
 `

File: test/unit/services/domain/test-reduction.js [hunk 6]
```diff
@@ -1,16 +1,77 @@
...
+
+  it('should raise a ValidationError if Notes exceed max length', function () {
+    var activeStartDate = moment().subtract(30, 'days').toDate()
+    var activeEndDate = moment().add(30, 'days').toDate()
+    expect(function () {
+      var longNote = getLongString()
+      new Reduction('1', '10',
+        [activeStartDate.getDate(), activeStartDate.getMonth() + 1, activeStartDate.getFullYear()],
+        [activeEndDate.getDate(), activeEndDate.getMonth() + 1, activeEndDate.getFullYear()], longNote)
+    }).to.throw(ValidationError)
```
Comment (caoimheb) [inline test/unit/services/domain/test-reduction.js]: Can we add regex to the throw check to make sure error is the correct one?

File: test/unit/services/helpers/test-reduction-helper.js [hunk 7]
```diff
@@ -1,20 +1,26 @@
...
+  [activeStartDate.getDate(), activeStartDate.getMonth() + 1, activeStartDate.getFullYear()],
+  [activeEndDate.getDate(), activeEndDate.getMonth() + 1, activeEndDate.getFullYear()], 'active note')
+var SCHEDULED_REDUCTION = new Reduction('2', '12',
+  [scheduledStartDate.getDate(), scheduledStartDate.getMonth() + 1, scheduledStartDate.getFullYear()],
+  [scheduledEndDate.getDate(), scheduledEndDate.getMonth() + 1, scheduledEndDate.getFullYear()], 'scheduled note')
+var ARCHIVE_REDUCTION = new Reduction('3', '13',
+  [archivedStartDate.getDate(), archivedStartDate.getMonth() + 1, archivedStartDate.getFullYear()],
+  [archivedEndDate.getDate(), archivedEndDate.getMonth() + 1, archivedEndDate.getFullYear()], 'archive note')
 
-var REDUCTIONS = [ACTIVE_REDUCTION, SCHEDULED_REDUCTION, ARCHIVE_REDUCTION, DELETED_REDUCTION]
```
Comment (caoimheb) [inline test/unit/services/helpers/test-reduction-helper.js]: Any reason `DELETED` was removed? Think it was included to illustrate that a deleted reduction will not get added to any array by the helper

File: test/unit/services/validators/test-field-validator.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+      var errors = errorHandler.get()
+      expect(errors).to.have.property(FIELD_NAME)
+    })
+
+    it('should throw error if data is an object', function () {
+      var errorHandler = ErrorHandler()
+      FieldValidator({}, FIELD_NAME, errorHandler)
+        .isRequired()
+      var errors = errorHandler.get()
+      expect(errors).to.equal(false)
```
Comment (caoimheb) [inline test/unit/services/validators/test-field-validator.js]: Is this checking that an error is thrown?

File: test/unit/services/validators/test-field-validator.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+      expect(errors).to.have.property(FIELD_NAME)
+    })
+
+    it('leading zeros: should return an error if passed a value interpreted as octal', function () {
+      var min = 1.0
+      var max = 99.0
+      FieldValidator('072', FIELD_NAME, errorHandler).isFloat(min, max)
+      var errors = errorHandler.get()
+      expect(errors).to.equal(false)
+    })
```
Comment (caoimheb) [inline test/unit/services/validators/test-field-validator.js]: Tests for `isLessThanLength`?

File: test/unit/services/validators/test-fieldset-validator.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+      this.fieldsetValidator.isFutureDate(FUTURE_DATE)
+      var errors = this.error.get()
+      expect(errors).to.equal(false)
+    })
+
+    it('should return the fieldsetValidator after being called to allow function chaining.', function () {
+      var result = this.fieldsetValidator.isFutureDate(dateFormatter.now())
+      expect(result).to.be.equal(this.fieldsetValidator)
+    })
+  })
```
Comment (caoimheb) [inline test/unit/services/validators/test-fieldset-validator.js]: Tests for other functions?

File: test/unit/services/validators/test-validation-error-messages.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    expect(message).to.equal(expected)
+  })
+
+  it('should return expected IsRange validation error message', function () {
+    var min = 1
+    var max = 99
+    const expected = `${fieldName} must be between ${min} and ${max}`
+    var message = ValidationErrorMessages.getIsRangeMessage(fieldName, { min: min, max: max })
+    expect(message).to.equal(expected)
+  })
```
Comment (caoimheb) [inline test/unit/services/validators/test-validation-error-messages.js]: Tests for other error messages?

File: app/routes/reductions.js [hunk 12]
```diff
@@ -189,27 +242,10 @@ module.exports = function (router) {
...
-    } else if ((reductionStartDate > currentDate) && (reductionEndDate > currentDate)) {
-      status = reductionStatusType.SCHEDULED
-    } else if ((reductionStartDate < currentDate) && (reductionEndDate < currentDate)) {
-      status = reductionStatusType.ARCHIVED
-    }
-
-    return status
+    var reductionStartDate = [ requestBody.red_start_day, requestBody.red_start_month, requestBody.red_start_year ]
+    var reductionEndDate = [ requestBody.red_end_day, requestBody.red_end_month, requestBody.red_end_year ]
+    var reasonId = !requestBody.reasonForReductionId ? 'select' : requestBody.reasonForReductionId
```
Comment (caoimheb) [inline app/routes/reductions.js]: Why not let the `undefined` pass through here?

File: app/services/domain/reduction.js [hunk 13]
```diff
@@ -1,11 +1,73 @@
...
+    this.reasonForReductionId = parseInt(this.reasonForReductionId)
+    this.hours = parseFloat(this.hours)
+    this.reductionStartDate = startDate.toDate()
+    this.reductionEndDate = endDate.toDate()
+  }
+
+  setReductionStatus () {
+    var currentDate = new Date()
+    var status = reductionStatusType.ACTIVE
+    if ((this.reductionStartDate < currentDate) && (this.reductionEndDate > currentDate)) {
```
[the remark this answers is not in the document]
Comment (tswann) [reply app/services/domain/reduction.js]: 👍

File: app/services/domain/workload-points.js (hunk 1 above)
Comment (tswann) [reply app/services/domain/workload-points.js]: This is based off the max values for the historical data. All points values are three digits max.

Contracted Hours is a question I've raised. Current known value is 37, but could possibly change in future.

File: test/unit/services/validators/test-field-validator.js (hunk 8 above)
Comment (tswann) [reply test/unit/services/validators/test-field-validator.js]: Nope - just that handler returns `false` if the internal collection is empty.

Comment (caoimheb) [reply test/unit/services/validators/test-field-validator.js]: So is the test description correct?

File: app/services/validators/field-validator.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    this.data = data
+    this.fieldName = fieldName
+    this.errors = errors
+  }
+
+  isRequired (specificMessage) {
+    var message = (!specificMessage) ? ERROR_MESSAGES.getIsRequiredMessage : specificMessage
+    if (!this.data) {
+      this.errors.add(this.fieldName, message)
+    } else if (this.data === 'select') {
```
Comment (caoimheb) [inline app/services/validators/field-validator.js]: Is this check needed anymore now the value isn't being set to `select` in the reduction route?

File: app/views/workload-points.html (hunk 2 above)
Comment (caoimheb) [reply app/views/workload-points.html]: ?

File: app/views/add-reduction.html [hunk 15]
```diff
@@ -17,78 +17,91 @@
...
-  {%if failureText %}
-  <p class="message message--failure">{{ failureText }}</p>
-  {% endif %}
+  <header class="heading-xlarge" style="margin-top:0; margin-bottom:0">{{ headerText }}</header>
 
   <p class="lede">{{ reductionText }}<span class="bold">{{ title }}</span></p>
   {% set first = true %}
   <form action="{{ formAction }}" method=POST id="reductionForm">
     <input type="hidden" name="reductionId" value="{{reduction.id}}"/>
     <div class="form-group" id="reduction_reason_selection">
```
Comment (andrew-js-wright) [inline app/views/add-reduction.html]: The IDs for the form elements should probably be updated to the value referred to by the error message link e.g. `reasonForReductionId`

Comment (andrew-js-wright) [reply app/views/add-reduction.html]: This is the case for each view.

File: app/views/workload-points.html (hunk 2 above)
Comment (tswann) [reply app/views/workload-points.html]: Missed this one! Will take a look 👍

File: app/routes/reductions.js [hunk 16]
```diff
@@ -36,25 +37,25 @@ module.exports = function (router) {
...
       throw new Error('Only available for offender manager')
     }
 
-    var failureText = fail ? 'Something went wrong. Please try again.' : null
     var getAddReductionsReferenceDataPromise = reductionsService.getAddReductionsRefData(id, organisationLevel)
 
     return getAddReductionsReferenceDataPromise
       .then(function (result) {
+        var errors = req.session.addReductionErrors
+        delete req.session.addReductionErrors
```
[the remark this answers is not in the document]
Comment (tswann) [reply app/routes/reductions.js]: 👍

File: app/services/domain/reduction.js [hunk 17]
```diff
@@ -1,11 +1,71 @@
...
+    this.reasonForReductionId = parseInt(this.reasonForReductionId)
+    this.hours = parseFloat(this.hours)
+    this.reductionStartDate = startDate.toDate()
+    this.reductionEndDate = endDate.toDate()
+  }
+
+  setReductionStatus () {
+    var currentDate = new Date()
+    var status = reductionStatusType.ACTIVE
+    if ((this.reductionStartDate > currentDate) && (this.reductionEndDate > currentDate)) {
```
[the remark this answers is not in the document]
Comment (tswann) [reply app/services/domain/reduction.js]: 👍

File: app/services/domain/reduction.js [hunk 18]
```diff
@@ -1,11 +1,71 @@
...
+      status = reductionStatusType.SCHEDULED
+    } else if ((this.reductionStartDate < currentDate) && (this.reductionEndDate < currentDate)) {
```
[the remark this answers is not in the document]
Comment (tswann) [reply app/services/domain/reduction.js]: 👍

File: test/unit/services/validators/test-field-validator.js (hunk 8 above)
Comment (tswann) [reply test/unit/services/validators/test-field-validator.js]: Ah, sorry my mistake. It's actually the isRequired Handler that's wrong! I've added a lodash check for empty objects to the check:
`_.isEmpty(this.data)`

File: app/views/add-reduction.html (hunk 15 above)
Comment (tswann) [reply app/views/add-reduction.html]: As long as name matches we're OK with the id being different

File: app/services/domain/workload-points.js [hunk 19]
```diff
@@ -0,0 +1,148 @@
+const ValidationError = require('../errors/validation-error')
+const ErrorHandler = require('../validators/error-handler')
+const FieldValidator = require('../validators/field-validator')
+
+class WorkloadPoints {
+  constructor (request) {
+    this.previousWpId = request.previousWpId
```
[the remark this answers is not in the document]
Comment (tswann) [reply app/services/domain/workload-points.js]: 👍

File: app/views/add-reduction.html (hunk 15 above)
Comment (andrew-js-wright) [reply app/views/add-reduction.html]: Ah, I didn't realise that. Thanks for the clarification.

**Pull request merged**